src/Npgsql/NpgsqlDataSourceBuilder.cs

```diff
@@ -66,17 +65,28 @@ public NpgsqlDataSource Build()
 
         if (ConnectionStringBuilder.Host!.Contains(","))
         {
+            if (ConnectionStringBuilder.TargetSessionAttributes is not null)
+                throw new InvalidOperationException(NpgsqlStrings.CannotSpecifyTargetSessionAttributes);
             if (ConnectionStringBuilder.Multiplexing)
                 throw new NotSupportedException("Multiplexing is not supported with multiple hosts");
             if (ConnectionStringBuilder.ReplicationMode != ReplicationMode.Off)
                 throw new NotSupportedException("Replication is not supported with multiple hosts");
-            return new MultiHostDataSource(ConnectionStringBuilder, connectionString, loggingConfiguration);
+
+            return new NpgsqlMultiHostDataSource(ConnectionStringBuilder, loggingConfiguration);
         }
 
         return ConnectionStringBuilder.Multiplexing
-            ? new MultiplexingDataSource(ConnectionStringBuilder, connectionString, loggingConfiguration)
+            ? new MultiplexingDataSource(ConnectionStringBuilder, loggingConfiguration)
             : ConnectionStringBuilder.Pooling
-                ? new PoolingDataSource(ConnectionStringBuilder, connectionString, loggingConfiguration)
-                : new UnpooledDataSource(ConnectionStringBuilder, connectionString, loggingConfiguration);
+                ? new PoolingDataSource(ConnectionStringBuilder, loggingConfiguration)
+                : new UnpooledDataSource(ConnectionStringBuilder, loggingConfiguration);
     }
+
+#pragma warning disable RS0016
+    /// <summary>
+    /// Builds and returns a <see cref="NpgsqlMultiHostDataSource" /> which is ready for use for load-balancing and failover scenarios.
+    /// </summary>
+    public NpgsqlMultiHostDataSource BuildMultiHost()
```

Comment:
Could/should? this be an extension method?

Reply:
Any particular reason? It's all built-in support, I don't anticipate us introducing many other types of data source types etc.

Reply:
Yeah, that's why it's a double question.

Reply:
I think it's OK... And we usually don't care about binary compat that much that we couldn't change it to an extension in a major version if we really wanted to...